substrate/frame/delegated-staking/src/types.rs

```diff
@@ -0,0 +1,308 @@
+#[derive(Default, Clone, Encode, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct AgentLedger<T: Config> {
+	/// Where the reward should be paid out.
+	pub payee: T::AccountId,
+	/// Sum of all delegated funds to this `Agent`.
+	#[codec(compact)]
+	pub total_delegated: BalanceOf<T>,
+	/// Funds that are withdrawn from core staking but not released to delegator/s. It is a subset
+	/// of `total_delegated` and can never be greater than it.
+	///
+	/// We need this register to ensure that the `Agent` does not bond funds from delegated
+	/// funds that are withdrawn and should be claimed by delegators.
+	#[codec(compact)]
+	pub unclaimed_withdrawals: BalanceOf<T>,
+	/// Slashes that are not yet applied. This affects the effective balance of the `Agent`.
+	#[codec(compact)]
+	pub pending_slash: BalanceOf<T>,
+}
+
+impl<T: Config> AgentLedger<T> {
+	/// Create a new instance of `AgentLedger`.
+	pub(crate) fn new(reward_destination: &T::AccountId) -> Self {
+		AgentLedger {
+			payee: reward_destination.clone(),
+			total_delegated: Zero::zero(),
+			unclaimed_withdrawals: Zero::zero(),
+			pending_slash: Zero::zero(),
+		}
+	}
+
+	/// Get `AgentLedger` from storage.
+	pub(crate) fn get(key: &T::AccountId) -> Option<Self> {
+		<Agents<T>>::get(key)
+	}
+
+	/// Save self to storage with the given key.
+	pub(crate) fn save(self, key: &T::AccountId) {
+		<Agents<T>>::insert(key, self)
+	}
+
+	/// Effective total balance of the `Agent`.
+	///
+	/// This takes into account any slashes reported to `Agent` but unapplied.
+	pub(crate) fn effective_balance(&self) -> BalanceOf<T> {
+		defensive_assert!(
+			self.total_delegated >= self.pending_slash,
+			"slash cannot be higher than actual balance of delegator"
+		);
+
+		// pending slash needs to be burned and cannot be used for stake.
+		self.total_delegated.saturating_sub(self.pending_slash)
+	}
+
+	/// Agent balance that can be staked/bonded in [`T::CoreStaking`].
+	pub(crate) fn stakeable_balance(&self) -> BalanceOf<T> {
+		self.effective_balance().saturating_sub(self.unclaimed_withdrawals)
+	}
+}
+
+/// Wrapper around `AgentLedger` to provide some helper functions to mutate the ledger.
+#[derive(Clone)]
+pub struct Agent<T: Config> {
+	/// storage key
+	pub key: T::AccountId,
+	/// storage value
+	pub ledger: AgentLedger<T>,
+}
+
+impl<T: Config> Agent<T> {
+	/// Get `Agent` from storage if it exists or return an error.
+	pub(crate) fn from(agent: &T::AccountId) -> Result<Agent<T>, DispatchError> {
+		let ledger = AgentLedger::<T>::get(agent).ok_or(Error::<T>::NotAgent)?;
+		Ok(Agent { key: agent.clone(), ledger })
+	}
+
+	/// Remove funds that are withdrawn from [Config::CoreStaking] but not claimed by a delegator.
+	///
+	/// Checked decrease of delegation amount from `total_delegated` and `unclaimed_withdrawals`
+	/// registers. Consumes self and returns a new instance of self if success.
+	pub(crate) fn remove_unclaimed_withdraw(
+		self,
```

Comment:
sth to keep in mind is that all FRAME dispatchables are not executing in a transactional layer, so if you take a mutable reference, and still return `Result<(), Err>`, and the return is `Err`, eventually any change to `self` will revert.

But if you find this better, I also get it.

Reply:
All frame dispatchables are executing in a tx layer right?

Didn't really get the point though?

Reply:
It was meant to encourage you to change more function signatures to `fn foo(&mut self) -> Result<(), Error>`.